File: FinalProject/app/src/main/java/com/example/finalproject/fragments/ShopFragment.java
package com.example.finalproject.fragments;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.SwitchCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.util.Log;
import android.util.Pair;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CompoundButton;
import android.widget.TextView;
import android.widget.Toast;

import com.example.finalproject.R;
import com.example.finalproject.items.Item;
import com.example.finalproject.service.Triplex;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.GenericTypeIndicator;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.HashMap;

public class ShopFragment extends Fragment {
    private static Item chosenItem;
    private static String user;
    private static int amount=1;
    private static boolean mode, auction;
    private static TextView amount_text;
    private static FragmentManager fm1;

    public ShopFragment() {
        auction=false;
    }

    public ShopFragment(boolean auction) {
        ShopFragment.auction = auction;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        fm1=getChildFragmentManager();
        return inflater.inflate(R.layout.fragment_shop, container, false);
    }

    @Override
    public void onViewCreated(@NonNull @NotNull View view, @Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        SwitchCompat sellBuy=getView().findViewById(R.id.sell_buy_switch);
        TextView modeText=getView().findViewById(R.id.shop_mode);
        amount_text=getView().findViewById(R.id.amount);
        Button sellBuyButton=getView().findViewById(R.id.sell_buy_button),
                back=getView().findViewById(R.id.shop_back_button);
        FloatingActionButton add=getView().findViewById(R.id.add_button),
                remove=getView().findViewById(R.id.remove_button);
        RecyclerView shopList=getView().findViewById(R.id.items_to_sell);
        shopList.setLayoutManager(new LinearLayoutManager(getContext()));
        ArrayList<Pair<Item, Integer>> data=new ArrayList<>();
        for (int i=0;i<MainActivity.shopList.size();i++)
            data.add(new Pair<>(MainActivity.shopList.get(i), 999));
        shopList.setAdapter(new ShopAdapter(MainActivity.player.getInventory()));
        if (!auction){
            sellBuy.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
                @Override
                public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                    mode=isChecked;
                    if (isChecked){
                        modeText.setText("Buy");
                        sellBuyButton.setText("Buy");
                        shopList.setAdapter(new ShopAdapter(data));
                    }
                    else {
                        modeText.setText("Sell");
                        sellBuyButton.setText("Sell");
                        shopList.setAdapter(new ShopAdapter(MainActivity.player.getInventory()));
                    }
                    amount=1;
                    amount_text.setText(amount+"");
                }
            });
            sellBuy.setChecked(true);
            sellBuy.setChecked(false);
            sellBuyButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (!mode){
                        if(!MainActivity.player.sellItem(new Pair<>(chosenItem, amount)))
                            Toast.makeText(getContext(), "You don't have enough items", Toast.LENGTH_SHORT).show();
                        else {
                            Toast.makeText(getContext(), "Sold successfully", Toast.LENGTH_SHORT).show();
                            shopList.setAdapter(new ShopAdapter(MainActivity.player.getInventory()));
                            FragmentManager fm=getChildFragmentManager();
                            FragmentTransaction fr=fm.beginTransaction();
                            if (fm.findFragmentById(R.id.characteristics3)!=null)
                                fr.remove(fm.findFragmentById(R.id.characteristics3));
                            fr.commit();
                        }
                    } else
                        if (!MainActivity.player.buyItem(new Pair<>(chosenItem, amount)))
                            Toast.makeText(getContext(), "You don't have enough gold", Toast.LENGTH_SHORT).show();
                        else Toast.makeText(getContext(), "Bought successfully", Toast.LENGTH_SHORT).show();
                }
            });
        }
        else {
            modeText.setText("Place");
            sellBuy.setText("Place/Buy");
            sellBuyButton.setText("Place");
            shopList.setAdapter(new ShopAdapter(MainActivity.player.getInventory()));
            sellBuy.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
                @Override
                public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                    mode=isChecked;
                    FragmentTransaction fr=fm1.beginTransaction();
                    if (fm1.findFragmentById(R.id.characteristics3) != null)
                        fr.remove(fm1.findFragmentById(R.id.characteristics3));
                    fr.commit();
                    if (isChecked){
                        modeText.setText("Buy");
                        sellBuyButton.setText("Buy");
                        shopList.setAdapter(new ShopAdapter(new ArrayList<>()));
                        DatabaseReference ref=FirebaseDatabase.getInstance().getReference("Auction");
                        GenericTypeIndicator<ArrayList<Triplex<Item, Integer, String>>> genericTypeIndicator
                                = new GenericTypeIndicator<ArrayList<Triplex<Item, Integer, String>>>() {};
                        ref.get().addOnCompleteListener(new OnCompleteListener<DataSnapshot>() {
                            @Override
                            public void onComplete(@NonNull @NotNull Task<DataSnapshot> task) {
                                if (task.isSuccessful())
                                    if (task.getResult().getValue(genericTypeIndicator)!=null)
                                        shopList.setAdapter(new ShopAdapter(task.getResult().getValue(genericTypeIndicator), true));
                            }
                        });
                    }
                    else {
                        modeText.setText("Place");
                        sellBuyButton.setText("Place");
                        shopList.setAdapter(new ShopAdapter(MainActivity.player.getInventory()));
                    }
                    amount_text.setText(amount+"");
                }
            });
            sellBuyButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (chosenItem!=null){
                        if (!mode)
                            if (amount>MainActivity.player.getAmountOfItemsInInventory(chosenItem))
                                Toast.makeText(getContext(), "You don't have enough items", Toast.LENGTH_SHORT).show();
                            else {
                                MainActivity.player.placeItemOnAuction(
                                        new Triplex<>(chosenItem, amount, MainActivity.player.getUser().getLogin()), shopList);
                                FragmentTransaction fr=fm1.beginTransaction();
                                if (fm1.findFragmentById(R.id.characteristics3) != null)
                                    fr.remove(fm1.findFragmentById(R.id.characteristics3));
                                fr.commit();
                            }
                            else{
                                MainActivity.player.buyItemOnAuction(amount,
                                        chosenItem.getName(), user, shopList, getContext());
                            amount_text.setText(amount+"");
                            FragmentTransaction fr=fm1.beginTransaction();
                            if (fm1.findFragmentById(R.id.characteristics3) != null)
                                fr.remove(fm1.findFragmentById(R.id.characteristics3));
                            fr.commit();
                            }
                    }
                    else Toast.makeText(getContext(), "Please, choose item first", Toast.LENGTH_SHORT).show();
                }
            });
        }
        add.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                amount++;
                amount_text.setText(amount+"");
            }
        });
        remove.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (amount>1){
                    amount--;
                    amount_text.setText(amount+"");
                }
            }
        });
        back.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                FragmentManager fm=getParentFragmentManager();
                FragmentTransaction fr=fm.beginTransaction();
                fr.remove(fm.findFragmentById(R.id.shop));
                if (fm.findFragmentById(R.id.map)!=null)
                    fr.remove(fm.findFragmentById(R.id.map));
                fr.add(R.id.map, new MapFragment(MainActivity.player.getMapNum()));
                fr.add(R.id.status, new StatusBarFragment());
                fr.add(R.id.menu, new MenuFragment());
                fr.commit();
            }
        });
    }

    public static class ShopAdapter extends RecyclerView.Adapter<ShopFragment.ShopAdapter.ViewHolder> {
        private ArrayList<Pair<Item, Integer>> data = new ArrayList<>();
        private ArrayList<Triplex<Item, Integer, String>> dataA=new ArrayList<>();

        public ShopAdapter(ArrayList<Pair<Item, Integer>> data) {
            this.data.addAll(data);
        }

        public ShopAdapter(ArrayList<Triplex<Item, Integer, String>> dataA, boolean auction){
            this.dataA.addAll(dataA);
        }

        private class ViewHolder extends RecyclerView.ViewHolder {
            TextView name;

            public ViewHolder(@NonNull @NotNull View itemView) {
                super(itemView);
                name = (TextView) itemView.findViewById(R.id.textView);
            }
        }

        @NonNull
        @NotNull
        @Override
        public ShopFragment.ShopAdapter.ViewHolder onCreateViewHolder(@NonNull @NotNull ViewGroup parent, int viewType) {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.inventory_item, parent, false);
            return new ViewHolder(view);
        }

        @Override
        public void onBindViewHolder(@NonNull @NotNull ShopFragment.ShopAdapter.ViewHolder holder, int position) {
            if (!auction){
                if (data.size()==0){
                    FragmentTransaction fr=fm1.beginTransaction();
                    if (fm1.findFragmentById(R.id.characteristics3) != null)
                        fr.remove(fm1.findFragmentById(R.id.characteristics3));
                    fr.commit();
                }
                else {
                    if (!mode)
                        holder.name.setText(">"+data.get(position).first.getName()+":"+data.get(position).second+"\n");
                    else holder.name.setText(">"+data.get(position).first.getName()+"\n");
                    holder.name.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            chosenItem=data.get(position).first;
                            amount=1;
                            amount_text.setText(amount+"");
                            FragmentTransaction fr=fm1.beginTransaction();
                            if (fm1.findFragmentById(R.id.characteristics3) != null)
                                fr.remove(fm1.findFragmentById(R.id.characteristics3));
                            fr.add(R.id.characteristics3, new ItemCharacteristicsFragment(data.get(position).first, !mode));
                            fr.commit();
                        }
                    });
                }
            } else {
                if (dataA.size()==0&&mode||data.size()==0&&!mode){
                    FragmentTransaction fr=fm1.beginTransaction();
                    if (fm1.findFragmentById(R.id.characteristics3) != null)
                        fr.remove(fm1.findFragmentById(R.id.characteristics3));
                    fr.commit();
                } else if (mode){
                    holder.name.setText(">"+dataA.get(position).first.getName()+":"+dataA.get(position).second+"\n"
                            +dataA.get(position).third+"\n");
                    holder.name.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            chosenItem=dataA.get(position).first;
                            user=dataA.get(position).third;
                            amount=1;
                            amount_text.setText(amount+"");
                            FragmentTransaction fr=fm1.beginTransaction();
                            if (fm1.findFragmentById(R.id.characteristics3) != null)
                                fr.remove(fm1.findFragmentById(R.id.characteristics3));
                            fr.add(R.id.characteristics3, new ItemCharacteristicsFragment(dataA.get(position).first, !mode));
                            fr.commit();
                        }
                    });
                }
                else {
                    holder.name.setText(">"+data.get(position).first.getName()+":"+data.get(position).second+"\n");
                    holder.name.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            chosenItem=data.get(position).first;
                            amount=1;
                            amount_text.setText(amount + "");
                            FragmentTransaction fr=fm1.beginTransaction();
                            if (fm1.findFragmentById(R.id.characteristics3) != null)
                                fr.remove(fm1.findFragmentById(R.id.characteristics3));
                            fr.add(R.id.characteristics3, new ItemCharacteristicsFragment(data.get(position).first, mode));
                            fr.commit();
                        }
                    });
                }
            }
        }

        @Override
        public int getItemCount() {
            ArrayList a= new ArrayList();
            if (auction)
                if (mode){
                 for (int i=0;i<dataA.size();i++){
                     if (dataA.get(i).second!=0)
                         a.add(dataA.get(i));
                 }
                 dataA.clear();
                 dataA.addAll(a);
                 return dataA.size();
                }
            return data.size();
        }
    }
}

File: FinalProject/app/src/main/java/com/example/finalproject/fragments/ShopFragment.kt
package com.example.finalproject.fragments

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.CompoundButton
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.widget.SwitchCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.finalproject.MainActivity.Companion.player
import com.example.finalproject.R
import com.example.finalproject.service.Statistics
import com.example.finalproject.service.classes.items.Item
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.GenericTypeIndicator

class ShopFragment(private val auction: Boolean = false) : Fragment(), View.OnClickListener,
    CompoundButton.OnCheckedChangeListener {

    private val inf: Int = 1e9.toInt()

    private var chosenItem: Item? = null
    private var user: String? = null
    private var amount = 1
    private var buyMode = false
    private val data = ArrayList<Pair<Int, Item>>()
    private lateinit var amountText: TextView
    private lateinit var modeText: TextView
    private lateinit var sellBuyButton: Button
    private lateinit var back: Button
    private lateinit var sellBuy: SwitchCompat
    private lateinit var add: FloatingActionButton
    private lateinit var remove: FloatingActionButton
    private lateinit var shopList: RecyclerView
    private lateinit var _childFragmentManager: FragmentManager

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_shop, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        _childFragmentManager = childFragmentManager
        sellBuy = requireView().findViewById(R.id.sell_buy_switch)
        modeText = requireView().findViewById(R.id.shop_mode)
        amountText = requireView().findViewById(R.id.amount)
        sellBuyButton = requireView().findViewById(R.id.sell_buy_button)
        back = requireView().findViewById(R.id.shop_back_button)
        add = requireView().findViewById(R.id.add_button)
        remove = requireView().findViewById(R.id.remove_button)
        shopList = requireView().findViewById(R.id.items_to_sell)
        shopList.layoutManager = LinearLayoutManager(context)
        for (i in Statistics.shopList)
            data.add(Pair(inf, i))
        shopList.adapter = ShopAdapter(player.inventory.inventory)
        if (!auction) {
            sellBuy.setOnCheckedChangeListener(this)
            sellBuy.isChecked = true
            sellBuy.isChecked = false
            sellBuyButton.setOnClickListener(this)
        } else {
            var text="Place"
            modeText.text = text
            sellBuyButton.text = text
            text="Place/Buy"
            sellBuy.text = text
            shopList.adapter = ShopAdapter(player.inventory.inventory)
            sellBuy.setOnCheckedChangeListener(this)
            sellBuyButton.setOnClickListener(this)
        }
        add.setOnClickListener(this)
        remove.setOnClickListener(this)
        back.setOnClickListener(this)
    }

    private inner class ShopAdapter : RecyclerView.Adapter<ShopAdapter.ViewHolder> {
        private val data = ArrayList<Pair<Int, Item>>()
        private val dataA = ArrayList<Triple<Item, Int, String>>()
        private var auction: Boolean

        constructor(data: ArrayList<Pair<Int, Item>>) {
            auction = false
            this.data.addAll(data)
        }

        constructor(dataA: ArrayList<Triple<Item, Int, String>>, auction: Boolean) {
            this.auction = true
            this.dataA.addAll(dataA)
        }

        private inner class ViewHolder(itemView: View) :
            RecyclerView.ViewHolder(itemView) {
            var name: TextView = itemView.findViewById<View>(R.id.textView) as TextView
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            val view =
                LayoutInflater.from(parent.context).inflate(R.layout.inventory_item, parent, false)
            return ViewHolder(view)
        }

        @SuppressLint("SetTextI18n")
        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            if (!auction) {
                if (data.size == 0) {
                    val fr = _childFragmentManager.beginTransaction()
                    _childFragmentManager.findFragmentById(R.id.characteristics3)
                        ?.let { fr.remove(it) }
                    fr.commit()
                } else {
                    if (!buyMode)
                        holder.name.text = ">${data[position].second.name}:${data[position].first}"
                    else holder.name.text = ">${data[position].second.name}"
                    holder.name.setOnClickListener {
                        chosenItem = data[position].second
                        amount = 1
                        amountText.text = amount.toString()
                        val fr = _childFragmentManager.beginTransaction()
                        _childFragmentManager.findFragmentById(R.id.characteristics3)
                            ?.let { it1 -> fr.remove(it1) }
                        fr.add(
                            R.id.characteristics3,
                            ItemCharacteristicsFragment(data[position].second, !buyMode)
                        )
                        fr.commit()
                    }
                }
            } else {
                if (dataA.size == 0 && buyMode || data.size == 0 && !buyMode) {
                    val fragmentTransaction = _childFragmentManager.beginTransaction()
                    _childFragmentManager.findFragmentById(R.id.characteristics3)
                        ?.let { fragmentTransaction.remove(it) }
                    fragmentTransaction.commit()
                } else if (buyMode) {
                    holder.name.text =">${dataA[position].first.name}:${dataA[position].second} ${dataA[position].third}"
                    holder.name.setOnClickListener {
                        chosenItem =
                            dataA[position].first
                        user =
                            dataA[position].third
                        amount = 1
                        amountText.text = amount.toString()
                        val fr = _childFragmentManager.beginTransaction()
                        _childFragmentManager.findFragmentById(R.id.characteristics3)
                            ?.let { it1 -> fr.remove(it1) }
                        fr.add(R.id.characteristics3,
                            ItemCharacteristicsFragment(dataA[position].first, !buyMode))
                        fr.commit()
                    }
                } else {
                    holder.name.text =">${data[position].second.name}:${data[position].first}"
                    holder.name.setOnClickListener {
                        chosenItem =
                            data[position].second
                        amount = 1
                        amountText.text = amount.toString()
                        val fr = _childFragmentManager.beginTransaction()
                        _childFragmentManager.findFragmentById(R.id.characteristics3)
                            ?.let { it1 -> fr.remove(it1) }
                        fr.add(R.id.characteristics3,
                            ItemCharacteristicsFragment(data[position].second, buyMode))
                        fr.commit()
                    }
                }
            }
        }

        override fun getItemCount(): Int {
            val a: ArrayList<Triple<Item, Int, String>> = ArrayList()
            if (auction)
                if (buyMode) {
                    for (i in dataA.indices) {
                        if (dataA[i].second != 0)
                            a.add(dataA[i])
                    }
                    dataA.clear()
                    dataA.addAll(a)
                    return dataA.size
                }
            return data.size
        }
    }

    override fun onClick(p0: View?) {
        if (p0==add)
            amountText.text = (++amount).toString()
        else if (p0==remove&&amount>1)
            amountText.text = (--amount).toString()
        else if (p0==back){
            val fragmentManager = parentFragmentManager
            val fragmentTransaction = fragmentManager.beginTransaction()
            fragmentManager.findFragmentById(R.id.shop)?.let { fragmentTransaction.remove(it) }
            fragmentManager.findFragmentById(R.id.map)?.let { fragmentTransaction.remove(it) }
            fragmentTransaction.add(R.id.map, MapFragment(player.mapNumber))
            fragmentTransaction.add(R.id.status, StatusBarFragment())
            fragmentTransaction.add(R.id.menu, MenuFragment())
            fragmentTransaction.commit()
        } else{
            if (auction) {
                if (p0==sellBuyButton){
                    if (chosenItem != null) {
                        if (!buyMode)
                            if (amount > player.inventory.quantity(chosenItem!!))
                                Toast.makeText(context, "You don't have enough items", Toast.LENGTH_SHORT).show()
                            else {
                                val fragmentTransaction = _childFragmentManager.beginTransaction()
                                _childFragmentManager.findFragmentById(R.id.characteristics3)
                                    ?.let { fragmentTransaction.remove(it) }
                                fragmentTransaction.commit()
                            }
                        else {
                            //player.buyItemOnAuction()
                            amountText.text = amount.toString()
                            val fragmentTransaction = _childFragmentManager.beginTransaction()
                            _childFragmentManager.findFragmentById(R.id.characteristics3)
                                ?.let { fragmentTransaction.remove(it) }
                            fragmentTransaction.commit()
                        }
                    } else Toast.makeText(context, "Please, choose item first", Toast.LENGTH_SHORT)
                        .show()
                }
            } else {
                if (p0 == sellBuyButton) {
                    if (chosenItem == null)
                        Toast.makeText(context, "Choose item first", Toast.LENGTH_SHORT).show()
                    else if (!buyMode) {
                        if (!player.sellItem(Pair(amount, chosenItem!!)))
                            Toast.makeText(
                                context,
                                "You don't have enough items", Toast.LENGTH_SHORT
                            ).show()
                        else {
                            Toast.makeText(context, "Sold successfully", Toast.LENGTH_SHORT).show()
                            shopList.adapter = ShopAdapter(player.inventory.inventory)
                            val fr = _childFragmentManager.beginTransaction()
                            _childFragmentManager.findFragmentById(R.id.characteristics3)?.let { it1 -> fr.remove(it1) }
                            fr.commit()
                        }
                    } else if (player.buyItem(Pair(amount, chosenItem!!)))
                        Toast.makeText(context, "You don't have enough gold", Toast.LENGTH_SHORT).show()
                    else
                        Toast.makeText(context, "Bought successfully", Toast.LENGTH_SHORT).show()
                }
            }
        }
    }

    override fun onCheckedChanged(p0: CompoundButton?, p1: Boolean) {
        if (auction) {
            if (p0==sellBuy){
                buyMode = p1
                val fr = _childFragmentManager.beginTransaction()
                _childFragmentManager.findFragmentById(R.id.characteristics3)?.let { fr.remove(it) }
                fr.commit()
                if (p1) {
                    val text="Buy"
                    modeText.text = text
                    sellBuyButton.text = text
                    shopList.adapter = ShopAdapter(ArrayList())
                    val ref = FirebaseDatabase.getInstance().getReference("Auction")
                    val genericTypeIndicator: GenericTypeIndicator<ArrayList<Triple<Item, Int, String>>> =
                        object : GenericTypeIndicator<ArrayList<Triple<Item, Int, String>>>() {}
                    ref.get().addOnCompleteListener { task ->
                        if (task.isSuccessful)
                            if (task.result.getValue(genericTypeIndicator) != null )
                                shopList.adapter = ShopAdapter(task.result.getValue(genericTypeIndicator)!!, true)
                    }
                } else {
                    val text="Place"
                    modeText.text = text
                    sellBuyButton.text = text
                    shopList.adapter = ShopAdapter(player.inventory.inventory)
                }
                amountText.text = amount.toString()
            }
        } else {
            if (p0 == sellBuy) {
                buyMode = p1
                if (buyMode) {
                    val text="Buy"
                    modeText.text = text
                    sellBuyButton.text = text
                    shopList.adapter = ShopAdapter(data)
                } else {
                    val text="Sell"
                    modeText.text = text
                    sellBuyButton.text = text
                    shopList.adapter = ShopAdapter(player.inventory.inventory)
                }
                amount=1
                amountText.text = amount.toString()
            }
        }
    }
}